### PR TITLE
fix(compiler-cli): improve the missingControlFlowDirective message

### PR DESCRIPTION
**Pull request by cexbrayat**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
The extended diagnostics about missing control flow directive was only mentioning that the `CommonModule` should be imported.
```
error NG8103: The `*ngFor` directive was used in the template, but the `CommonModule` was not imported. Please make sure that the `CommonModule` is included into the `@Component.imports` array of this component.
```

## What is the new behavior?
Now that the control flow directives are available as standalone, the message mentions that directive itself can be imported. It also refers to the import that should be used, which can be tricky to figure out for `ngFor`:

```
error NG8103: The `*ngFor` directive was used in the template, but the `NgForOf` directive or the `CommonModule` were not imported. Please make sure that the `NgForOf` directive or the `CommonModule` are included in the `@Component.imports` array of this component.
```

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_control_flow_directive/index.ts [hunk 1]
```diff
@@ -65,8 +65,8 @@ class MissingControlFlowDirectiveCheck extends
 
     const sourceSpan = controlFlowAttr.keySpan || controlFlowAttr.sourceSpan;
     const errorMessage = `The \`*${controlFlowAttr.name}\` directive was used in the template, ` +
-        `but the \`CommonModule\` was not imported. Please make sure that the \`CommonModule\` ` +
-        `is included into the \`@Component.imports\` array of this component.`;
+        `but the directive or the \`CommonModule\` were not imported. Please make sure that the directive or the \`CommonModule\` ` +
+        `are included in the \`@Component.imports\` array of this component.`;
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_control_flow_directive/index.ts:69]: Can we make `KNOWN_CONTROL_FLOW_DIRECTIVES` a map from template attribute names to directive names, allowing this to actually name the directive to import? Especially for `ngFor` it's nice to know that you need to import `NgForOf`, as there is no `NgFor`.

Comment (cexbrayat) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_control_flow_directive/index.ts]: Great idea, I updated the PR, let me know what you think.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_control_flow_directive/index.ts [hunk 2]
```diff
@@ -64,9 +68,11 @@ class MissingControlFlowDirectiveCheck extends
     }
 
     const sourceSpan = controlFlowAttr.keySpan || controlFlowAttr.sourceSpan;
+    const correspondingImport = KNOWN_CONTROL_FLOW_DIRECTIVES.get(controlFlowAttr.name);
     const errorMessage = `The \`*${controlFlowAttr.name}\` directive was used in the template, ` +
-        `but the \`CommonModule\` was not imported. Please make sure that the \`CommonModule\` ` +
-        `is included into the \`@Component.imports\` array of this component.`;
+        `but the \`${correspondingImport}\` directive or the \`CommonModule\` were not imported. ` +
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_control_flow_directive/index.ts:73]: I'm not sure if this should be `were imported` as you changed it, or if it should stay `was imported` (only a single candidate needs to be imported, not multiple).

**Review by JoostK — APPROVED** [review]
LGTM; I'm also adding @AndrewKushnir as reviewer to get his thoughts.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_control_flow_directive/index.ts (hunk 2 above)
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_control_flow_directive/index.ts:73]:
```suggestion
        `but neither the \`${correspondingImport}\` directive nor the \`CommonModule\` were not imported. ` +
```

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_control_flow_directive/index.ts [hunk 3]
```diff
@@ -64,9 +68,11 @@ class MissingControlFlowDirectiveCheck extends
...
+        `Please make sure that the \`${correspondingImport}\` directive or the \`CommonModule\` ` +
+        `are included in the \`@Component.imports\` array of this component.`;
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_control_flow_directive/index.ts:75]:
```suggestion
        `Please make sure that either the \`${correspondingImport}\` directive or the \`CommonModule\` ` +
        `are included in the \`@Component.imports\` array of this component.`;
```

File: packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/missing_control_flow_directive/missing_control_flow_directive_spec.ts [hunk 4]
```diff
@@ -17,10 +17,11 @@ import {factory as missingControlFlowDirectiveCheck} from '../../../checks/missi
 import {ExtendedTemplateCheckerImpl} from '../../../src/extended_template_checker';
 
 const KNOWN_CONTROL_FLOW_DIRECTIVES = ['ngIf', 'ngFor', 'ngSwitchCase', 'ngSwitchDefault'];
+const CORRESPONDING_IMPORTS = ['NgIf', 'NgForOf', 'NgSwitchCase', 'NgSwitchDefault'];
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/missing_control_flow_directive/missing_control_flow_directive_spec.ts:20]: I think it'd be great to have the same structure as the one we have in `packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_control_flow_directive/index.ts` and may be even import that Map here (not sure why I didn't do that when implementing this...).

**Review by AndrewKushnir — COMMENTED** [review]
@cexbrayat the code looks good, thanks for updating the message 👍 

There are also some checks in `packages/core/src/render3/instructions/element_validation.ts` related to "known control flow directives". Could you please apply a similar update there for consistency?

Thank you.

**Comment by cexbrayat** [comment]
@AndrewKushnir @JoostK Thanks for the reviews. I updated the PR with your feedback.

@AndrewKushnir I'll open a following PR to update the validation in core 👍

**Review by AndrewKushnir — APPROVED** [review]
👍

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit 7882746913c48c623ec9abca7257f90b2fd4773c.

**Pull request closed without merging**